src/ring_theory/nilpotent.lean

```diff
@@ -0,0 +1,103 @@
+/-
+Copyright (c) 2021 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import data.nat.choose.sum
+
+/-!
+# Nilpotent elements
+
+## Main definitions
+
+  * `is_nilpotent`
+  * `is_nilpotent_neg_iff`
+  * `commute.is_nilpotent_add`
+  * `commute.is_nilpotent_mul_left`
+  * `commute.is_nilpotent_mul_right`
+  * `commute.is_nilpotent_sub`
+
+-/
+
+universes u
+
+variables {R : Type u} {x y : R}
+
+/-- An element is said to be nilpotent if some natural-number-power of it equals zero.
+
+Note that we require only the bare minimum assumptions for the definition to make sense. Even
+`monoid_with_zero` is too strong since nilpotency is important in the study of rings that are only
+power-associative. -/
+def is_nilpotent [has_zero R] [has_pow R ℕ] (x : R) : Prop := ∃ (n : ℕ), x^n = 0
+
+lemma monoid_with_zero.pow_eq_zero_of_le
+  [monoid_with_zero R] {n m : ℕ} (hn : n ≤ m) (hx : x^n = 0) : x^m = 0 :=
+by rw [← nat.sub_add_cancel hn, pow_add, hx, mul_zero]
+
+lemma is_nilpotent.zero [monoid_with_zero R] : is_nilpotent (0 : R) := ⟨1, pow_one 0⟩
+
+lemma is_nilpotent.neg [ring R] (h : is_nilpotent x) : is_nilpotent (-x) :=
+begin
+  obtain ⟨n, hn⟩ := h,
+  use n,
+  rw [neg_pow, hn, mul_zero],
+end
+
+lemma is_nilpotent_neg_iff [ring R] : is_nilpotent (-x) ↔ is_nilpotent x :=
+⟨λ h, neg_neg x ▸ h.neg, λ h, h.neg⟩
+
+lemma eq_zero_of_is_nilpotent [monoid_with_zero R] [no_zero_divisors R]
```

Comment:
```suggestion
@[simp] lemma is_nilpotent_iff_eq_zero [monoid_with_zero R] [no_zero_divisors R]
```
would be a better lema where, the reverse direction is just `is_nilpotent.zero`

Reply:
```suggestion
lemma is_nilpotent.eq_zero [monoid_with_zero R] [no_zero_divisors R]
```
for dot notation

Reply:
Both good points. Done.